### PR TITLE
JBTokenStore tests

### PR DESCRIPTION
**Pull request by dhyon**
image

### CODE REVIEW

File: test/jb_token_store/change_for.test.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import jbOperatoreStore from '../../artifacts/contracts/JBOperatorStore.sol/JBOperatorStore.json';
+import jbProjects from '../../artifacts/contracts/JBProjects.sol/JBProjects.json';
+import jbToken from '../../artifacts/contracts/JBToken.sol/JBToken.json';
+import { Contract } from 'ethers';
+
+describe('JBTokenStore::changeFor(...)', function () {
+  const PROJECT_ID = 2;
+
+  async function setup() {
+    let [deployer, ...addrs] = await ethers.getSigners();
```
Comment (joax) [inline test/jb_token_store/change_for.test.js:17]: Add "owner" and "newOwner" to clean code of unnecessary variable uses.

File: test/jb_token_store/change_for.test.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    await mockJbDirectory.mock.controllerOf.withArgs(PROJECT_ID).returns(controller.address);
+
+    // Issue the initial token and grab a reference to it.
+    await jbTokenStore.connect(controller).issueFor(PROJECT_ID, 'TestTokenDAO', 'TEST');
+    let initialTokenAddr = await jbTokenStore.connect(controller).tokenOf(PROJECT_ID);
+    let initialToken = new Contract(initialTokenAddr, jbToken.abi);
+
+    // Change to a new token.
+    let newOwner = addrs[2];
+    let newTokenName = 'NewTokenDAO';
```
Comment (joax) [inline test/jb_token_store/change_for.test.js:50]: This should be a const NEW_TOKEN_NAME at the beginning of the test.

File: test/jb_token_store/change_for.test.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    let newTokenSymbol = 'NEW';
+    let newToken = await deployToken(newTokenName, newTokenSymbol);
```
Comment (joax) [inline test/jb_token_store/change_for.test.js]: Same as above, NEW_TOKEN_SYMBOL. These are used all over the place.

File: test/jb_token_store/change_for.test.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  async function deployToken(name, symbol) {
+    let jbTokenFactory = await ethers.getContractFactory('JBToken');
+    return await jbTokenFactory.deploy(name, symbol);
+  }
+
+  it('Should change tokens and emit event if caller is controller', async function () {
+    const { addrs, mockJbDirectory, jbTokenStore } = await setup();
+    let controller = addrs[1];
```
Comment (joax) [inline test/jb_token_store/change_for.test.js:39]: Define this on the setup()

File: test/jb_token_store/balance_of.test.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+import { deployMockContract } from '@ethereum-waffle/mock-contract';
+
+import jbDirectory from '../../artifacts/contracts/JBDirectory.sol/JBDirectory.json';
+import jbOperatoreStore from '../../artifacts/contracts/JBOperatorStore.sol/JBOperatorStore.json';
+import jbProjects from '../../artifacts/contracts/JBProjects.sol/JBProjects.json';
+
+describe('JBTokenStore::balanceOf(...)', function () {
+  const PROJECT_ID = 2;
+  const name = 'TestTokenDAO';
```
Comment (joax) [inline test/jb_token_store/balance_of.test.js:12]: const TOKEN_NAME = 'TestTokenDAO';
const TOKEN_SYMBOL = 'TEST';

File: contracts/JBTokenStore.sol [hunk 6]
```diff
@@ -264,6 +264,9 @@ contract JBTokenStore is JBControllerUtility, JBOperatable, IJBTokenStore {
     uint256 _amount,
     bool _preferClaimedTokens
   ) external override onlyController(_projectId) {
+    // An amount must be specified.
+    require(_amount > 0, '0x22: NO_OP');
```
Comment (dhyon) [inline contracts/JBTokenStore.sol:268]: this is the only change I figured we needed in this contract

**Review by joax — CHANGES_REQUESTED** [review]
Left some comments to keep consistency among the tests. Take a look at the comments and I can review again.

- push the vars as `const` to the beginning of the test
- pre-define `tokenOwner` and `newTokenOwner` at `setup()` to avoid variables in the test cases
- pre-define `tokenName`, `newTokenName`, `tokenSymbol` and others at the beginning of the tests as `const` to avoid variables creation on the test cases if reused by multiple cases
- use the variable naming for `const` with caps (i.e. TOKEN_NAME and NEW_TOKEN_NAME) when defining the `const` at the beginning.

File: test/jb_token_store/change_for.test.js (hunk 1 above)
Comment (dhyon) [reply test/jb_token_store/change_for.test.js:17]: `owner` isn't used in this test suite and `newOwner` is used in one test so I think it should be fine to define it once in the local scope

Comment (dhyon) [reply test/jb_token_store/change_for.test.js:17]: This definitely applies to other files though

**Review by joax — COMMENTED** [review]
All good with me. @odd-amphora.

File: test/jb_token_store/balance_of.test.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    const { newHolder, controller, mockJbDirectory, jbTokenStore } = await setup();
+
+    await mockJbDirectory.mock.controllerOf.withArgs(PROJECT_ID).returns(controller.address);
+
+    await jbTokenStore.connect(controller).issueFor(PROJECT_ID, TOKEN_NAME, TOKEN_SYMBOL);
+
+    // Mint unclaimed tokens
+    const numTokens = 20;
+    await jbTokenStore.connect(controller).mintFor(newHolder.address, PROJECT_ID, numTokens, false);
+
```
Comment (odd-amphora) [inline test/jb_token_store/balance_of.test.js:47]: Can we add an expectation here before minting claimed tokens?

File: test/jb_token_store/balance_of.test.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  it('Should return token balance for holder', async function () {
+    const { newHolder, controller, mockJbDirectory, jbTokenStore } = await setup();
+
+    await mockJbDirectory.mock.controllerOf.withArgs(PROJECT_ID).returns(controller.address);
+
+    await jbTokenStore.connect(controller).issueFor(PROJECT_ID, TOKEN_NAME, TOKEN_SYMBOL);
+
+    // Mint unclaimed tokens
+    const numTokens = 20;
+    await jbTokenStore.connect(controller).mintFor(newHolder.address, PROJECT_ID, numTokens, false);
```
Comment (odd-amphora) [inline test/jb_token_store/balance_of.test.js:46]: consider using parameter comments for the boolean args /*preferClaimedTokens=*/

File: test/jb_token_store/transfer_to.test.js [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+
+  it(`Can't transfer unclaimed tokens when recipient and holder are the same`, async function () {
+    const { controller, holder, mockJbOperatorStore, jbTokenStore, TRANSFER_INDEX } = await setup();
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(controller.address, holder.address, PROJECT_ID, TRANSFER_INDEX)
+      .returns(true);
+
+    await expect(
+      jbTokenStore.connect(controller).transferTo(holder.address, holder.address, PROJECT_ID, 1),
```
Comment (odd-amphora) [inline test/jb_token_store/transfer_to.test.js:105]: Consider using parameter comment for `/*amount=*/` when it's not clear.

File: test/jb_token_store/total_supply_of.test.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    const controller = addrs[0];
+
+    await mockJbDirectory.mock.controllerOf.withArgs(PROJECT_ID).returns(controller.address);
+
+    await jbTokenStore.connect(controller).issueFor(PROJECT_ID, TOKEN_NAME, TOKEN_SYMBOL);
+
+    // Mint unclaimed tokens
+    const newHolder = addrs[1];
+    const numTokens = 20;
+    await jbTokenStore.connect(controller).mintFor(newHolder.address, PROJECT_ID, numTokens, false);
```
Comment (odd-amphora) [inline test/jb_token_store/total_supply_of.test.js:47]: Consider using param comment for last param.

File: test/jb_token_store/should_require_claiming_for.test.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      REQUIRE_CLAIM_INDEX,
+    } = await setup();
+
+    await mockJbProjects.mock.ownerOf.withArgs(PROJECT_ID).returns(controller.address);
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(controller.address, holder.address, PROJECT_ID, REQUIRE_CLAIM_INDEX)
+      .returns(true);
+
+    await expect(
+      jbTokenStore.connect(controller).shouldRequireClaimingFor(PROJECT_ID, true),
```
Comment (odd-amphora) [inline test/jb_token_store/should_require_claiming_for.test.js:95]: consider using param comment for last param

**Review by odd-amphora — APPROVED** [review]

File: contracts/JBTokenStore.sol (hunk 6 above)
Comment (dhyon) [reply contracts/JBTokenStore.sol:268]: @mejango 🗒️

**Review by mejango — APPROVED** [review]
yes!

**Pull request merged**